**About Book Crossing Dataset**<br>

This dataset has been compiled by Cai-Nicolas Ziegler in 2004, and it comprises of three tables for users, books and ratings. Explicit ratings are expressed on a scale from 1-10 (higher values denoting higher appreciation) and implicit rating is expressed by 0.

Reference: http://www2.informatik.uni-freiburg.de/~cziegler/BX/ 

**Objective**

This project entails building a Book Recommender System for users based on user-based and item-based collaborative filtering approaches.

#### Execute the below cell to load the datasets

In [463]:
!pip install scikit-surprise
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [464]:
#Loading data
books = pd.read_csv("books.csv", sep=";", error_bad_lines=False, encoding="latin-1")
books.columns = ['ISBN', 'bookTitle', 'bookAuthor', 'yearOfPublication', 'publisher', 'imageUrlS', 'imageUrlM', 'imageUrlL']

users = pd.read_csv('users.csv', sep=';', error_bad_lines=False, encoding="latin-1")
users.columns = ['userID', 'Location', 'Age']

ratings = pd.read_csv('ratings.csv', sep=';', error_bad_lines=False, encoding="latin-1")
ratings.columns = ['userID', 'ISBN', 'bookRating']

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'


### Check no.of records and features given in each dataset

In [465]:
books.shape

(271360, 8)

In books dataset we have 271360 records and 8 fields with it.

In [466]:
users.shape

(278858, 3)

In users dataset we have 278858 records and 3 fields with it.

In [467]:
ratings.shape

(1149780, 3)

In books dataset we have 1149780 records and 3 fields with it.

## Exploring books dataset

In [468]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  \
0    Mark P. O. Morford              2002     Oxford University Press   
1  Richard Bruce Wright              2001       HarperFlamingo Canada   
2          Carlo D'Este              1991             HarperPerennial   
3      Gina Bari Kolata              1999        Farrar Straus Giroux   
4       E. J. W. Barber              1999  W. W. Norton &amp; Company   

                                           imageUrlS  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlM  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlL  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

### Drop last three columns containing image URLs which will not be required for analysis

In [469]:
books=books.drop(["imageUrlS","imageUrlM","imageUrlL"],axis=1)

In [470]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  
0    Mark P. O. Morford              2002     Oxford University Press  
1  Richard Bruce Wright              2001       HarperFlamingo Canada  
2          Carlo D'Este              1991             HarperPerennial  
3      Gina Bari Kolata              1999        Farrar Straus Giroux  
4       E. J. W. Barber              1999  W. W. Norton &amp; Company

**yearOfPublication**

### Check unique values of yearOfPublication


In [471]:
books['yearOfPublication'].unique

<bound method Series.unique of 0         2002
1         2001
2         1991
3         1999
4         1999
5         1991
6         2000
7         1993
8         1996
9         2002
10        1988
11        2000
12        2004
13        1999
14        1998
15        1998
16        1999
17        1994
18        1999
19        1994
20        1999
21        2001
22        2003
23        2000
24        1999
25        1998
26        2004
27        1997
28        2000
29        1996
          ... 
271330    1990
271331    1997
271332    1993
271333    1994
271334    2001
271335    1989
271336    1995
271337    2002
271338    1997
271339    2003
271340    2003
271341    1998
271342    2001
271343    1998
271344    1999
271345    1999
271346    1978
271347    2000
271348    2003
271349    2004
271350    2002
271351    2001
271352    2004
271353    1975
271354    1993
271355    1988
271356    1991
271357    2004
271358    1996
271359    2000
Name: yearOfPublication, Length: 271360, dtype: object

In [472]:
print("Number of unique values in yearOfPublication is : " ,books['yearOfPublication'].nunique())

Number of unique values in yearOfPublication is :  202


In [473]:
books.yearOfPublication.value_counts()

2002         13903
2001         13715
1999         13414
2000         13373
1998         12116
2003         11610
1997         11494
1996         10687
1995         10259
1994          8857
1993          7920
1992          7390
1991          6926
1990          6394
1989          5825
1988          5545
1987          4761
2004          4629
1986          4258
1999          4017
1985          3912
2000          3859
2002          3724
1998          3650
2001          3644
1984          3631
0             3570
1997          3396
1996          3343
1983          3297
             ...  
1904             1
2020             1
Gallimard        1
1917             1
1900             1
2037             1
1806             1
2024             1
1376             1
1926             1
1914             1
1927             1
1924             1
1378             1
2008             1
1897             1
2012             1
1944             1
2011             1
2021             1
2026             1
2038        

In [474]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 5 columns):
ISBN                 271360 non-null object
bookTitle            271360 non-null object
bookAuthor           271359 non-null object
yearOfPublication    271360 non-null object
publisher            271358 non-null object
dtypes: object(5)
memory usage: 10.4+ MB


As it can be seen from above that there are some incorrect entries in this field. It looks like Publisher names 'DK Publishing Inc' and 'Gallimard' have been incorrectly loaded as yearOfPublication in dataset due to some errors in csv file.


Also some of the entries are strings and same years have been entered as numbers in some places. We will try to fix these things in the coming questions.

### Check the rows having 'DK Publishing Inc' as yearOfPublication

In [475]:
books.loc[books['yearOfPublication'] == 'DK Publishing Inc']

ISBN                                          bookTitle  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

       bookAuthor  yearOfPublication  \
209538       2000  DK Publishing Inc   
221678       2000  DK Publishing Inc   

                                                publisher  
209538  http://images.amazon.com/images/P/078946697X.0...  
221678  http://images.amazon.com/images/P/0789466953.0...

In [476]:
##Also checking for rows having 'Gallimard' as yearOfPublication
books.loc[books['yearOfPublication'] == 'Gallimard']

ISBN                                          bookTitle  \
220731  2070426769  Peuple du ciel, suivi de 'Les Bergers\";Jean-M...   

       bookAuthor yearOfPublication  \
220731       2003         Gallimard   

                                                publisher  
220731  http://images.amazon.com/images/P/2070426769.0...

### Drop the rows having `'DK Publishing Inc'` and `'Gallimard'` as `yearOfPublication`

In [477]:
books=books.drop(books.loc[books['yearOfPublication']== 'DK Publishing Inc'].index) 
books=books.drop(books.loc[books['yearOfPublication']== 'Gallimard'].index)

In [478]:
books['yearOfPublication'].count()

271357

So now 3 rows got deleted that were related to 'DK Publishing Inc' and 'Gallimard' as yearOfPublication

### Change the datatype of yearOfPublication to 'int'

In [479]:
books['yearOfPublication']=books['yearOfPublication'].astype(int)

In [480]:
books.dtypes

ISBN                 object
bookTitle            object
bookAuthor           object
yearOfPublication     int32
publisher            object
dtype: object

### Drop NaNs in `'publisher'` column

In [481]:
## Checking shape before drop
books.shape

(271357, 5)

In [482]:
books=books[books['publisher'].notnull()]

In [483]:
## Checking shape after drop
books.shape

(271355, 5)

## Exploring Users dataset

In [484]:
print(users.shape)
users.head()

(278858, 3)


userID                            Location   Age
0       1                  nyc, new york, usa   NaN
1       2           stockton, california, usa  18.0
2       3     moscow, yukon territory, russia   NaN
3       4           porto, v.n.gaia, portugal  17.0
4       5  farnborough, hants, united kingdom   NaN

### Get all unique values in ascending order for column `Age`

In [485]:
print(sorted(users['Age'].unique()))

[nan, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0, 31.0, 32.0, 33.0, 34.0, 35.0, 36.0, 37.0, 38.0, 39.0, 40.0, 41.0, 42.0, 43.0, 44.0, 45.0, 46.0, 47.0, 48.0, 49.0, 50.0, 51.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 58.0, 59.0, 60.0, 61.0, 62.0, 63.0, 64.0, 65.0, 66.0, 67.0, 68.0, 69.0, 70.0, 71.0, 72.0, 73.0, 74.0, 75.0, 76.0, 77.0, 78.0, 79.0, 80.0, 81.0, 82.0, 83.0, 84.0, 85.0, 86.0, 87.0, 88.0, 89.0, 90.0, 91.0, 92.0, 93.0, 94.0, 95.0, 96.0, 97.0, 98.0, 99.0, 100.0, 101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0, 108.0, 109.0, 110.0, 111.0, 113.0, 114.0, 115.0, 116.0, 118.0, 119.0, 123.0, 124.0, 127.0, 128.0, 132.0, 133.0, 136.0, 137.0, 138.0, 140.0, 141.0, 143.0, 146.0, 147.0, 148.0, 151.0, 152.0, 156.0, 157.0, 159.0, 162.0, 168.0, 172.0, 175.0, 183.0, 186.0, 189.0, 199.0, 200.0, 201.0, 204.0, 207.0, 208.0, 209.0, 210.0, 212.0, 219.0, 220.0, 223.0, 226.0

Age column has some invalid entries like nan, 0 and very high values like 100 and above

### Values below 5 and above 90 do not make much sense for our book rating case...hence replace these by NaNs

In [486]:
users['Age'].values[users['Age'] > 90]  = np.nan
users['Age'].values[users['Age'] < 5]  = np.nan

### Replace null values in column `Age` with mean

In [487]:
users["Age"].fillna(users["Age"].mean(), inplace = True) 

### Change the datatype of `Age` to `int`

In [488]:
users['Age']=users['Age'].astype(int)

In [489]:
print(sorted(users.Age.unique()))

[5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90]


## Exploring the Ratings Dataset

### check the shape

In [490]:
ratings.shape

(1149780, 3)

In [491]:
books.shape

(271355, 5)

In [492]:
n_users = users.shape[0]
n_books = books.shape[0]

In [493]:
ratings.head(5)

userID        ISBN  bookRating
0  276725  034545104X           0
1  276726  0155061224           5
2  276727  0446520802           0
3  276729  052165615X           3
4  276729  0521795028           6

In [494]:
users.tail(5)

userID                            Location  Age
278853  278854               portland, oregon, usa   34
278854  278855  tacoma, washington, united kingdom   50
278855  278856           brampton, ontario, canada   34
278856  278857           knoxville, tennessee, usa   34
278857  278858                dublin, n/a, ireland   34

### Ratings dataset should have books only which exist in our books dataset. Drop the remaining rows

In [495]:
ratings.drop(ratings[~ratings['ISBN'].isin(books['ISBN'])].index, inplace = True) 

In [496]:
##Checking shape after drop
ratings.shape

(1031130, 3)

### Ratings dataset should have ratings from users which exist in users dataset. Drop the remaining rows

In [497]:
ratings.drop(ratings[~ratings['userID'].isin(users['userID'])].index, inplace = True) 

In [498]:
ratings.shape

(1031130, 3)

### Consider only ratings from 1-10 and leave 0s in column `bookRating`

In [499]:
ratings.drop(ratings[(ratings['bookRating']<1) | (ratings['bookRating']>10)].index,inplace=True)

In [500]:
ratings.shape

(383839, 3)

### Find out which rating has been given highest number of times

In [501]:
ratings['bookRating'].value_counts()

8     91804
10    71225
7     66401
9     60776
5     45355
6     31687
4      7617
3      5118
2      2375
1      1481
Name: bookRating, dtype: int64

SO from Above result we can say that rating 8 was given maximum time

### **Collaborative Filtering Based Recommendation Systems**

### For more accurate results only consider users who have rated atleast 100 books

In [502]:
ratings['userID'].value_counts().tail(10)

109745    1
195803    1
179411    1
154831    1
152782    1
148684    1
124088    1
105655    1
103606    1
211175    1
Name: userID, dtype: int64

In [503]:
# Get the count of each value
value_counts = ratings['userID'].value_counts()


to_remove = value_counts[value_counts < 100].index


ratings = ratings[~ratings.userID.isin(to_remove)]

### Generating ratings matrix from explicit ratings


#### Note: since NaNs cannot be handled by training algorithms, replace these by 0, which indicates absence of ratings

In [522]:
from scipy.sparse import csr_matrix
# pivot ratings into movie features
df_rating_features = ratings.pivot(index='userID',columns='ISBN',values='bookRating').fillna(0)
# convert dataframe of movie features to scipy sparse matrix
mat_movie_features = csr_matrix(df_rating_features.values)
df_rating_features.head(10)

ISBN    0000913154  0001046438  000104687X  0001047213  0001047973  \
userID                                                               
2033           0.0         0.0         0.0         0.0         0.0   
2110           0.0         0.0         0.0         0.0         0.0   
2276           0.0         0.0         0.0         0.0         0.0   
4017           0.0         0.0         0.0         0.0         0.0   
4385           0.0         0.0         0.0         0.0         0.0   
5582           0.0         0.0         0.0         0.0         0.0   
6242           0.0         0.0         0.0         0.0         0.0   
6251           0.0         0.0         0.0         0.0         0.0   
6543           0.0         0.0         0.0         0.0         0.0   
6575           0.0         0.0         0.0         0.0         0.0   

ISBN    000104799X  0001048082  0001053736  0001053744  0001055607  ...  \
userID                                                              ...   
2033           0.0         0.0         0.0         0.0         0.0  ...   
2110           0.0         0.0         0.0         0.0         0.0  ...   
2276           0.0         0.0         0.0         0.0         0.0  ...   
4017           0.0         0.0         0.0         0.0         0.0  ...   
4385           0.0         0.0         0.0         0.0         0.0  ...   
5582           0.0         0.0         0.0         0.0         0.0  ...   
6242           0.0         0.0         0.0         0.0         0.0  ...   
6251           0.0         0.0         0.0         0.0         0.0  ...   
6543           0.0         0.0         0.0         0.0         0.0  ...   
6575           0.0         0.0         0.0         0.0         0.0  ...   

ISBN    B000092Q0A  B00009EF82  B00009NDAN  B0000DYXID  B0000T6KHI  \
userID                                                               
2033           0.0         0.0         0.0         0.0         0.0   
2110           0.0         0.0         0.0         0.0         0.0   
2276           0.0         0.0         0.0         0.0         0.0   
4017           0.0         0.0         0.0         0.0         0.0   
4385           0.0         0.0         0.0         0.0         0.0   
5582           0.0         0.0         0.0         0.0         0.0   
6242           0.0         0.0         0.0         0.0         0.0   
6251           0.0         0.0         0.0         0.0         0.0   
6543           0.0         0.0         0.0         0.0         0.0   
6575           0.0         0.0         0.0         0.0         0.0   

ISBN    B0000VZEJQ  B0000X8HIE  B00013AX9E  B0001I1KOG  B000234N3A  
userID                                                              
2033           0.0         0.0         0.0         0.0         0.0  
2110           0.0         0.0         0.0         0.0         0.0  
2276           0.0         0.0         0.0         0.0         0.0  
4017           0.0         0.0         0.0         0.0         0.0  
4385           0.0         0.0         0.0         0.0         0.0  
5582           0.0         0.0         0.0         0.0         0.0  
6242           0.0         0.0         0.0         0.0         0.0  
6251           0.0         0.0         0.0         0.0         0.0  
6543           0.0         0.0         0.0         0.0         0.0  
6575           0.0         0.0         0.0         0.0         0.0  

[10 rows x 66572 columns]

### Generate the predicted ratings using SVD with no.of singular values to be 50

In [523]:
from scipy.sparse.linalg import svds
U, sigma, Vt = svds(df_rating_features, k = 50)
sigma

array([147.92121613, 149.3438051 , 150.07400599, 152.20116297,
       152.87416391, 154.61308307, 154.80093432, 155.95760177,
       158.05646578, 159.21079484, 159.81670657, 162.01963916,
       162.77851768, 163.33054635, 166.02489324, 166.8162391 ,
       168.04972004, 170.77485167, 171.01325686, 173.29428498,
       174.57624968, 176.65724713, 178.61913749, 180.29517222,
       182.25079063, 184.10706957, 187.61687534, 189.75276623,
       190.96966388, 195.14643609, 199.83133018, 201.70083339,
       202.18713912, 203.48697581, 207.26449173, 209.92986988,
       213.23598777, 216.88280493, 224.26954726, 231.66186197,
       235.67095629, 249.9581775 , 252.02866425, 261.24756904,
       267.98197504, 281.0120779 , 293.69539562, 379.58327277,
       634.72875357, 680.30978318])

In [524]:
sigma = np.diag(sigma)
sigma

array([[147.92121613,   0.        ,   0.        , ...,   0.        ,
          0.        ,   0.        ],
       [  0.        , 149.3438051 ,   0.        , ...,   0.        ,
          0.        ,   0.        ],
       [  0.        ,   0.        , 150.07400599, ...,   0.        ,
          0.        ,   0.        ],
       ...,
       [  0.        ,   0.        ,   0.        , ..., 379.58327277,
          0.        ,   0.        ],
       [  0.        ,   0.        ,   0.        , ...,   0.        ,
        634.72875357,   0.        ],
       [  0.        ,   0.        ,   0.        , ...,   0.        ,
          0.        , 680.30978318]])

In [527]:
all_users_predicted_ratings = np.dot(np.dot(U, sigma), Vt)
preds_df = pd.DataFrame(all_users_predicted_ratings, columns = df_rating_features.columns)
preds_df.head()

ISBN  0000913154  0001046438  000104687X  0001047213  0001047973  000104799X  \
0       0.025341   -0.002146   -0.001431   -0.002146   -0.002146    0.002971   
1      -0.010012   -0.003669   -0.002446   -0.003669   -0.003669    0.001075   
2      -0.015054   -0.015457   -0.010304   -0.015457   -0.015457    0.007281   
3      -0.021499    0.035602    0.023735    0.035602    0.035602    0.030307   
4       0.002077   -0.007965   -0.005310   -0.007965   -0.007965    0.002947   

ISBN  0001048082  0001053736  0001053744  0001055607  ...  B000092Q0A  \
0      -0.003920    0.007035    0.007035    0.012316  ...    0.000180   
1       0.001440   -0.003500   -0.003500    0.001612  ...   -0.000363   
2      -0.014033    0.011941    0.011941    0.011796  ...   -0.000455   
3       0.024215   -0.001053   -0.001053    0.067579  ...    0.002971   
4       0.003057    0.000231    0.000231    0.006080  ...    0.002120   

ISBN  B00009EF82  B00009NDAN  B0000DYXID  B0000T6KHI  B0000VZEJQ  B0000X8HIE  \
0       0.000226    0.042081   -0.016804   -0.080028    0.004746    0.028314   
1       0.000403    0.008142    0.001104   -0.029224    0.000999    0.002363   
2       0.001907    0.047982    0.005737    0.117859    0.006945    0.003119   
3       0.009912    0.086248   -0.008818    0.016154    0.028848   -0.000125   
4       0.001597   -0.012181    0.009420    0.673459    0.002591   -0.008229   

ISBN  B00013AX9E  B0001I1KOG  B000234N3A  
0       0.000120   -0.001693    0.067503  
1      -0.000242    0.000029   -0.013059  
2      -0.000304    0.009009   -0.057692  
3       0.001981    0.031201   -0.046664  
4       0.001413    0.004918    0.047773  

[5 rows x 66572 columns]

### Take a particular user_id

### Lets find the recommendations for user with id `2110`

#### Note: Execute the below cells to get the variables loaded

In [528]:
userID = 2110

In [529]:
user_id = 2 #2nd row in ratings matrix and predicted matrix

### Get the predicted ratings for userID `2110` and sort them in descending order

In [532]:
sorted_user_predictions = preds_df.loc[user_id-1].sort_values(ascending = False)
sorted_user_predictions.head()

ISBN
059035342X    0.682444
0345370775    0.368946
0345384911    0.333624
043935806X    0.333209
044021145X    0.329336
Name: 1, dtype: float64

In [533]:
preds_df=preds_df.sort_values(userID,ascending=False)
preds_df[userID].head()

KeyError: 2110

### Create a dataframe with name `user_data` containing userID `2110` explicitly interacted books

In [534]:
user_data = ratings[ratings.userID == (userID)]

In [535]:
user_data.head()

userID        ISBN  bookRating
14448    2110  0060987529           7
14449    2110  0064472779           8
14450    2110  0140022651          10
14452    2110  0142302163           8
14453    2110  0151008116           5

In [536]:
user_data.shape

(103, 3)

### Combine the user_data and and corresponding book data(`book_data`) in a single dataframe with name `user_full_info`

In [537]:
book_data=books[books['ISBN'].isin(user_data['ISBN'])] 

In [538]:
book_data.shape

(103, 5)

In [539]:
book_data.head()

ISBN                                          bookTitle  \
246   0151008116                                         Life of Pi   
904   015216250X  So You Want to Be a Wizard: The First Book in ...   
1000  0064472779                                  All-American Girl   
1302  0345307674                     Return of the Jedi (Star Wars)   
1472  0671527215                  Hitchhikers's Guide to the Galaxy   

         bookAuthor  yearOfPublication           publisher  
246     Yann Martel               2002            Harcourt  
904     Diane Duane               2001  Magic Carpet Books  
1000      Meg Cabot               2003        HarperTrophy  
1302     James Kahn               1983       Del Rey Books  
1472  Douglas Adams               1984              Pocket

In [540]:
user_full_info = (user_data.merge(book_data, how = 'left', left_on = 'ISBN', right_on = 'ISBN').
                 sort_values(['bookRating'], ascending = False))

In [541]:
user_full_info.head()

userID        ISBN  bookRating  \
76    2110  067166865X          10   
52    2110  0590109715          10   
64    2110  0590629786          10   
63    2110  0590629778          10   
61    2110  059046678X          10   

                                            bookTitle           bookAuthor  \
76  STAR TREK YESTERDAY'S SON (Star Trek: The Orig...         A.C. Crispin   
52  The Andalite Chronicles (Elfangor's Journey, A...  Katherine Applegate   
64                      The Visitor (Animorphs, No 2)      K. A. Applegate   
63                     The Invasion (Animorphs, No 1)      K. A. Applegate   
61                                       The Yearbook       Peter Lerangis   

    yearOfPublication   publisher  
76               1988  Audioworks  
52               1997       Apple  
64               1996  Scholastic  
63               1996  Scholastic  
61               1994  Scholastic

### Get top 10 recommendations for above given userID from the books not already rated by that user

In [544]:
num_recommendations = False
user_row_number = user_id - 1
sorted_user_predictions = preds_df.loc[user_row_number].sort_values(ascending = False)
rec= (books[~books['ISBN'].isin(user_full_info['ISBN'])].
                      merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left',
                           left_on = 'ISBN',
                           right_on = 'ISBN').
                      rename(columns = {user_row_number: 'Predictions'}).
                      sort_values('Predictions', ascending = False).
                      iloc[:10, :-1])

In [547]:
print("Top 10 recommendations are :" )
rec

Top 10 recommendations are :


ISBN                                          bookTitle  \
1192  0345370775                                      Jurassic Park   
6184  0345384911                                       Crystal Line   
5458  043935806X  Harry Potter and the Order of the Phoenix (Boo...   
455   044021145X                                           The Firm   
2031  0451151259                                 Eyes of the Dragon   
5383  0439139597       Harry Potter and the Goblet of Fire (Book 4)   
3413  0439064872   Harry Potter and the Chamber of Secrets (Book 2)   
976   0380759497                 Xanth 15: The Color of Her Panties   
2435  0345353145                                             Sphere   
6048  0451167317                                      The Dark Half   

            bookAuthor  yearOfPublication                     publisher  
1192  Michael Crichton               1999              Ballantine Books  
6184    Anne McCaffrey               1993                 Del Rey Books  
5458     J. K. Rowling               2003                    Scholastic  
455       John Grisham               1992  Bantam Dell Publishing Group  
2031      Stephen King               1988           Penguin Putnam~mass  
5383     J. K. Rowling               2000                    Scholastic  
3413     J. K. Rowling               2000                    Scholastic  
976      Piers Anthony               1992                           Eos  
2435  MICHAEL CRICHTON               1988              Ballantine Books  
6048      Stephen King               1994                   Signet Book